File: android/src/main/java/com/rebeloid/unity_mediation/AdMethodChannelCaller.java
package com.rebeloid.unity_mediation;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.MethodChannel;

public abstract class AdMethodChannelCaller {
    private final BinaryMessenger binaryMessenger;
    private final Map<String, MethodChannel> adUnitChannels;

    protected AdMethodChannelCaller(BinaryMessenger binaryMessenger, Map<String, MethodChannel> adUnitChannels) {
        this.binaryMessenger = binaryMessenger;
        this.adUnitChannels = adUnitChannels;
    }

    protected void invokeMethod(String methodName, String adUnitId) {
        invokeMethod(methodName, adUnitId, new HashMap<>());
    }

    protected void invokeMethod(String methodName, String adUnitId, String errorCode, String errorMessage) {
        Map<String, String> arguments = new HashMap<>();
        arguments.put(UnityMediationConstants.ERROR_CODE_PARAMETER, errorCode);
        arguments.put(UnityMediationConstants.ERROR_MESSAGE_PARAMETER, errorMessage);
        invokeMethod(methodName, adUnitId, arguments);
    }

    protected void invokeMethod(String methodName, String adUnitId, Map<String, String> arguments) {
        arguments.put(UnityMediationConstants.AD_UNIT_ID_PARAMETER, adUnitId);
        MethodChannel channel = adUnitChannels.get(adUnitId);
        if (channel == null) {
            channel = new MethodChannel(binaryMessenger, UnityMediationConstants.VIDEO_AD_CHANNEL + "_" + adUnitId);
            adUnitChannels.put(adUnitId, channel);
        }
        channel.invokeMethod(methodName, arguments);
    }
}

File: ios/Classes/AdMethodChannelCaller.swift
class AdMethodChannelCaller : NSObject {
    var adUnitChannels: [String: FlutterMethodChannel];
    let messenger: FlutterBinaryMessenger
    
    init(messenger: FlutterBinaryMessenger, adUnitChannels: [String: FlutterMethodChannel]) {
        self.messenger = messenger
        self.adUnitChannels = adUnitChannels
    }
    
    func invokeMethod(_ methodName: String, _ adUnitId: String) {
        invokeMethod(methodName, adUnitId, arguments: [:])
    }

    func invokeMethod(_ methodName: String, _ adUnitId: String, errorCode: String, errorMessage: String){
        var arguments: [String:String] = [:]
        arguments[UnityMediationConstants.ERROR_CODE_PARAMETER]=errorCode
        arguments[UnityMediationConstants.ERROR_MESSAGE_PARAMETER]=errorMessage
        invokeMethod(methodName, adUnitId, arguments: arguments)
    }

    func invokeMethod(_ methodName: String, _ adUnitId: String, arguments: [String:String]){
        var args = arguments;
        args[UnityMediationConstants.AD_UNIT_ID_PARAMETER]=adUnitId
        var channel = adUnitChannels[adUnitId]
        if (channel == nil) {
            channel = FlutterMethodChannel(name: UnityMediationConstants.VIDEO_AD_CHANNEL + "_" + adUnitId, binaryMessenger: messenger)
            adUnitChannels[adUnitId] = channel
        }
        channel?.invokeMethod(methodName, arguments: args)
    }
}
